Fix errors and warnings building swift/Sema on Windows using MSVC

hughbe opened the pull request:
- Unused variables
- Fix a really hard-to-fix assertion that was horrible to find and fix :))
- other misc changes

jrose-apple commented on lib/Sema/ConstraintSystem.h:
```diff
@@ -1730,7 +1730,11 @@ class ConstraintSystem {
+                                            __alignof(T));
+#else
                                             __alignof__(T));
+#endif
```
Does C++11 `alignof` do the right thing in all cases here?

jrose-apple commented on lib/Sema/TypeCheckProtocol.cpp:
```diff
@@ -412,15 +418,27 @@ namespace {
+      // but not with MSVC.
+      // This causes an assertion compiling Swift code with an MSVC swift.exe
+      // Therefore, we  manually invoke getSyntheticEnvironment() before
+      // invoking takeRequirementToSyntheticMap()
```
Eh, I don't think you even need this comment. We consider it bad style to depend on left-to-right-ness even when compiling with Clang. Thank you for finding and fixing this!

jrose-apple commented on lib/Sema/ConstraintGraph.cpp:
```diff
@@ -27,7 +27,9 @@
 
+#if !defined(_MSC_VER)
 #pragma mark Graph construction/destruction
+#endif
```
Same comments about pragma mark: if we have to put this nonsense around it every time, it might not be pulling its weight. @DougGregor, @rudkx, @xedin?

rudkx replied:
@hughbe Can you just globally disable the "unknown pragma" warning in the build using "-wd#"?

rudkx commented:
@hughbe My personal preference would be to squash the two later commits into the first since they are just small tweaks addressing review feedback with the original commit.

rudkx commented on lib/Sema/CSApply.cpp:
```diff
@@ -114,7 +114,7 @@ Type Solution::computeSubstitutions(
                           ConformanceCheckFlags::Used));
-    (void)isOpenedAnyObject;
+    (void)&isOpenedAnyObject;
     assert((conformance ||
```
Out of curiosity, what diagnostic are you getting from VC++ that requires taking the address here?

hughbe replied:
Function missing parameter call list or something like that

hughbe commented:
@rudkx thanks for taking a look. I've removed all ifdefs for pragma mark, and squashed the commits as you asked. I can't trigger tests yet, so could you if we've got to that point?

rudkx commented on include/swift/Sema/TypeCheckRequest.h:
```diff
@@ -97,6 +98,9 @@ class TypeCheckRequest {
 #include "swift/Sema/TypeCheckRequestKinds.def"
     }
+
+    // Work around MSVC warning: not all control paths return a value
```
Nit: I'm not sure the comment is really needed. I suppose someone in theory might attempt to remove the llvm_unreachable() and do so if they don't get a warning on a particular platform, but that seems unlikely.

rudkx commented on include/swift/Sema/TypeCheckRequest.h:
```diff
@@ -97,6 +98,9 @@ class TypeCheckRequest {
     }
+
+    // Work around MSVC warning: not all control paths return a value
+    llvm_unreachable("All switch cases are covered");
```
Another nit: I understand where you're coming from with the particular text here (and in the other places you added llvm_unreachable) since it does appear to be currently covered, but generally we tend to word these as something like "Unhandled <kind,type,etc.> in switch!" so that if the unreachable is ever hit the message is clear.

rudkx commented on lib/Sema/Constraint.cpp:
```diff
@@ -324,8 +327,15 @@ void Constraint::dump(ConstraintSystem *CS) const {
+#if defined(_MSC_VER)
+#pragma warning(pop)
+#endif
 }
```
I'm really curious why this is only hitting here, and only with VC++. Do you know if clang transitively decides not to warn here because it's being called from another deprecated function?

hughbe replied:
So here' the error:  `warning C4996: 'swift::constraints::Constraint::dump': only for use within the debugger`

It comes from a `LLVM_ATTRIBUTE_DEPRECATED` use here: https://github.com/apple/swift/blob/33c322fd226b0345b3771e7a8875f284777d286a/include/swift/AST/AnyFunctionRef.h#L163

> decides not to warn here because it's being called from another deprecated function?

That would surprise me, as, following up the chain of function calls, we would get at least one deprecation warning if this were the behaviour - otherwise all of this would be dead code. I can't find any `#pragma clang diagnostic` disables for `deprecated-declarations`, but maybe its a CMake thing?

That said, I'm really not sure, sorry

rudkx replied:
Yeah it's not clear to me why the difference exists here. I sometimes hit the deprecation warnings when I add my own calls to dump() in code that I am working on, so I don't know why clang wouldn't emit something here as well, but that's not really your problem.

rudkx commented:
@swift-ci Please smoke test and merge

hughbe commented:
Strange... tests have passed but this wasn't automerged
